File: lclang/src/main/java/lclang/LCBaseExecutor.java
package lclang;

import lclang.exceptions.LCLangRuntimeException;
import lclang.exceptions.LCLangVariableNotFoundException;
import lclang.libs.lang.classes.LCClass;

import java.util.HashMap;

public class LCBaseExecutor {
    public final LCRootExecutor root;
    public final HashMap<String, LCClass> variables = new HashMap<>();

    public LCBaseExecutor(LCRootExecutor root) {
        this.root = root;
    }

    public LCBaseExecutor(LCBaseExecutor executor, boolean importVariables) {
        this(executor.root);
        if(importVariables) variables.putAll(executor.variables);
    }

    public Link getLink(Caller c, String name) throws LCLangRuntimeException {
        
        Link value;
        if(variables.containsKey(name))
            value = variables.get(name).createLink();
        else {
            Link link = root.getLink(c, name);
            if(link!=null) value = link;
            else value = new Link(caller -> {
                throw new LCLangVariableNotFoundException(name, caller);
            }, Link.State.NOTHING);
        }

        value.set = (caller, clazz) -> variables.put(name, clazz);

        return value;
    }
}

File: lclang/src/main/kotlin/lclang/LCBaseExecutor.kt
package lclang

import lclang.exceptions.VariableNotFoundException
import lclang.types.Types

open class LCBaseExecutor(val root: LCRootExecutor) {
    constructor(executor: LCBaseExecutor): this(executor.root)
    val variables = HashMap<String, Value>()

    fun getLink(name: String): Value {
        val value = variables[name] ?: root.globals[name] ?:
        Value(Types.UNDEFINED, {
            throw VariableNotFoundException(name, it)
        })

        value.set = { _, newValue ->
            variables[name] = newValue
        }

        return value
    }

     //fun visitLambda(ctx: lclangParser.LambdaContext): Value = LambdaMethod(this, ctx).asValue()
}